File: Python3/1043.py
class Node:
    def __init__(self,data):
        self.data = data
        self.left = None
        self.right = None

def insert(root,data):
    if root is None:
        root = Node(data)
        return root
    if data < root.data:
        root.left = insert(root.left,data)
    else:
        root.right = insert(root.right,data)
    return root

def Pre_order(root):
    if root is None:
        return
    pre.append(root.data)
    Pre_order(root.left)
    Pre_order(root.right)

def Pre_order_mirror(root):
    if root is None:
        return
    pre_mirror.append(root.data)
    Pre_order_mirror(root.right)
    Pre_order_mirror(root.left)


def show(arr):
    for i in range(len(arr)):
        print("%d"%arr[i],end='')
        if i != len(arr)-1:
            print(" ",end='')
def Post_order(root):
    if root is None:
        return
    Post_order(root.left)
    Post_order(root.right)
    post.append(root.data)

def Post_order_mirror(root):
    if root is None:
        return
    Post_order_mirror(root.right)
    Post_order_mirror(root.left)
    post_mirror.append(root.data)
pre = []
pre_mirror=[]
post = []
post_mirror = []
n = int(input())
origin = list(map(int,input().split()))
root = Node(origin[0])

for each in origin[1:]:
    root = insert(root,each)
Pre_order(root)
# print(pre)
if pre == origin:
    print("YES")
    Post_order(root)
    show(post)
else:
    Pre_order_mirror(root)
    if pre_mirror == origin:
        print("YES")
        Post_order_mirror(root)
        show(post_mirror)
    else:
        print("NO")

File: c++/1043.cpp
//
//  main.cpp
//  test
//
//  Created by 李寻欢 on 2019/9/26.
//  Copyright © 2019 李寻欢. All rights reserved.
//
#include <stdio.h>
#include <iostream>
#include <string>
#include <map>
#include <cmath>
#include <algorithm>
#include <stack>
#include <vector>
#include <queue>
#include <unordered_map>
#include <string.h>
using namespace std;


typedef struct node{
    int data;
    node* left;
    node* right;
}Node,*TreeNode;

int n,data;
vector<int> v;
vector<int> pre;
vector<int> pre_m;
vector<int> post;
TreeNode insert(TreeNode root,int data){
    if (!root) {
        root = new Node;
        root->data = data;
        root->left = root->right = NULL;
        return root;
    }
    if(data < root->data)
        root->left = insert(root->left, data);
    else
        root->right = insert(root->right, data);
    return root;
}
vector<int> v1;
void Preorder(TreeNode root){
    if(!root) return;
    pre.push_back(root->data);
    Preorder(root->left);
    Preorder(root->right);
}
int space = 0;
void Post(TreeNode root){
    if(!root) return;
    Post(root->left);
    Post(root->right);
    space++;
    printf("%d",root->data);
    if(space != n) printf(" ");
}

void Post_mirror(TreeNode root){
    if(!root) return;
    Post_mirror(root->right);
    Post_mirror(root->left);
    space++;
    printf("%d",root->data);
    if(space != n) printf(" ");
}

void Pre_mirror(TreeNode root){
    if(!root) return;
    pre_m.push_back(root->data);
    Pre_mirror(root->right);
    Pre_mirror(root->left);
}

int main(){
    scanf("%d",&n);
    TreeNode root = NULL;
    for(int i=0;i<n;++i){
        scanf("%d",&data);
        v.push_back(data);
        root = insert(root, data);
    }
    bool flag1 = true;
    Preorder(root);
    for(int i=0;i<n;++i){
        if(pre[i] != v[i]){
            flag1 = false;
            break;
        }
    }
    if(flag1){
        printf("YES\n");
        Post(root);
        return 0;
    }else{
        bool flag2 = true;
        Pre_mirror(root);
        for(int i=0;i<n;++i){
            if(pre_m[i] != v[i]){
                flag2 = false;
                break;
            }
        }
        if(flag2){
            printf("YES\n");
            Post_mirror(root);
        }else{
            printf("NO\n");
        }
    }
    
}
